Loki crds (#3936)
grafana/agent: docs/sources/flow/reference/components/loki.rules.kubernetes.md

---
title: loki.rules.kubernetes
labels:
  stage: beta
---

# loki.rules.kubernetes

{{< docs/shared lookup="flow/stability/beta.md" source="agent" >}}

`loki.rules.kubernetes` discovers `PrometheusRule` Kubernetes resources and
loads them into a Loki instance.

> * Multiple `loki.rules.kubernetes` components can be specified by giving them
>   different labels.

Reviewer: ```suggestion
* You can specify multiple `loki.rules.kubernetes` components by giving them different labels.
```

  `PrometheusRule` resources considered during reconciliation.
* Compatible with the Ruler APIs of Grafana Loki, Grafana Cloud, and Grafana Enterprise Metrics.
* Compatible with the `PrometheusRule` CRD from the [prometheus-operator][].
* This component accesses the Kubernetes REST API from [within a Pod][].

{{% admonition type="note" %}}
This component requires [Role-based access control (RBAC)][] to be set up

> in Kubernetes for the Agent to access it via the Kubernetes REST API.

Reviewer: ```suggestion
in Kubernetes for {{< param "PRODUCT_ROOT_NAME" >}} to access it via the Kubernetes REST API.
```

> {{% /admonition %}}

Reviewer: ```suggestion

Role-based access control (RBAC)]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/
{{% /admonition %}}
```

We need to move the link definition inside the admonition shortcode for the rendering to work correctly.

> > **NOTE**: This component requires [Role-based access control (RBAC)][] to be setup
> > in Kubernetes in order for the Agent to access it via the Kubernetes REST API.
> > For an example RBAC configuration please click [here](#example).

Reviewer: ```suggestion
{{% admonition type="note" %}}
This component requires [Role-based access control (RBAC)][] to be set up
in Kubernetes for the Agent to access it via the Kubernetes REST API.
{{% /admonition %}}
```

[Kubernetes label selectors]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
[prometheus-operator]: https://prometheus-operator.dev/

> [within a Pod]: https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/
> [Role-based access control (RBAC)]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/

Reviewer: ```suggestion
[within a Pod]: https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/
```

## Usage

```river
loki.rules.kubernetes "LABEL" {
  address = LOKI_RULER_URL
}
```

## Arguments

`loki.rules.kubernetes` supports the following arguments:

Name                     | Type       | Description                                              | Default | Required
-------------------------|------------|----------------------------------------------------------|---------|---------
`address`                | `string`   | URL of the Loki ruler.                                  |         | yes
`tenant_id`              | `string`   | Loki tenant ID.                                         |         | no
`use_legacy_routes`      | `bool`     | Whether to use deprecated ruler API endpoints.           | false   | no
`sync_interval`          | `duration` | Amount of time between reconciliations with Loki.       | "30s"   | no

> `loki_namespace_prefix` | `string`   | Prefix used to differentiate multiple agent deployments. | "agent" | no

Reviewer: ```suggestion
`loki_namespace_prefix` | `string`   | Prefix used to differentiate multiple {{< param "PRODUCT_ROOT_NAME" >}} deployments. | "agent" | no
```

`bearer_token`           | `secret`   | Bearer token to authenticate with.                       |         | no
`bearer_token_file`      | `string`   | File containing a bearer token to authenticate with.     |         | no
`proxy_url`              | `string`   | HTTP proxy to proxy requests through.                    |         | no
`follow_redirects`       | `bool`     | Whether redirects returned by the server should be followed. | `true` | no
`enable_http2`           | `bool`     | Whether HTTP2 is supported for requests.                 | `true`  | no

>  At most one of the following can be provided:

Reviewer: ```suggestion
 At most, one of the following can be provided:
 
```

 - [`bearer_token` argument](#arguments).
 - [`bearer_token_file` argument](#arguments).
 - [`basic_auth` block][basic_auth].
 - [`authorization` block][authorization].
 - [`oauth2` block][oauth2].

 [arguments]: #arguments

If no `tenant_id` is provided, the component assumes that the Loki instance at
`address` is running in single-tenant mode and no `X-Scope-OrgID` header is sent.

The `sync_interval` argument determines how often Loki's ruler API is accessed

> to reload the current state of rules. Interaction with the Kubernetes API works

Reviewer: ```suggestion
to reload the current state. Interaction with the Kubernetes API works
```

differently. Updates are processed as events from the Kubernetes API server
according to the informer pattern.

You can use the `loki_namespace_prefix` argument to separate the rules managed

> by multiple agent deployments across your infrastructure. You should set the prefix to a

Reviewer: ```suggestion
by multiple {{< param "PRODUCT_ROOT_NAME" >}} deployments across your infrastructure. You should set the prefix to a
```

> The `loki_namespace_prefix` argument can be used to separate the rules managed
> by multiple agent deployments across your infrastructure. It should be set to a
> unique value for each deployment.

Reviewer: ```suggestion
You can use the `loki_namespace_prefix` argument to separate the rules managed
by multiple agent deployments across your infrastructure. You should set the prefix to a
unique value for each deployment.
```

`loki_rules_events_retried_total`            | `counter`   | Number of events that were retried, partitioned by event type.
`loki_rules_client_request_duration_seconds` | `histogram` | Duration of requests to the Loki API.

## Example

This example creates a `loki.rules.kubernetes` component that loads discovered
rules to a local Loki instance under the `team-a` tenant. Only namespaces and
rules with the `agent` label set to `yes` are included.

```river
loki.rules.kubernetes "local" {
    address = "loki:3100"
    tenant_id = "team-a"

    rule_namespace_selector {
        match_labels = {
            agent = "yes",
        }
    }

    rule_selector {
        match_labels = {
            agent = "yes",
        }
    }
}
```

This example creates a `loki.rules.kubernetes` component that loads discovered
rules to Grafana Cloud.

```river
loki.rules.kubernetes "default" {
    address = "GRAFANA_CLOUD_URL"
    basic_auth {
        username = "GRAFANA_CLOUD_USER"
        password = "GRAFANA_CLOUD_API_KEY"
        // Alternatively, load the password from a file:
        // password_file = "GRAFANA_CLOUD_API_KEY_PATH"
    }
}
```

> The following example is an RBAC configuration for Kubernetes. It authorizes the Agent to query the Kubernetes REST API:

Reviewer: ```suggestion
The following example is an RBAC configuration for Kubernetes. It authorizes {{< param "PRODUCT_ROOT_NAME" >}} to query the Kubernetes REST API:
```

---
title: loki.rules.kubernetes
labels:
  stage: beta
---

# loki.rules.kubernetes

{{< docs/shared lookup="flow/stability/beta.md" source="agent" >}}

`loki.rules.kubernetes` discovers `PrometheusRule` Kubernetes resources and
loads them into a Loki instance.

* You can specify multiple `loki.rules.kubernetes` components by giving them different labels.
* [Kubernetes label selectors][] can be used to limit the `Namespace` and
  `PrometheusRule` resources considered during reconciliation.
* Compatible with the Ruler APIs of Grafana Loki, Grafana Cloud, and Grafana Enterprise Metrics.
* Compatible with the `PrometheusRule` CRD from the [prometheus-operator][].
* This component accesses the Kubernetes REST API from [within a Pod][].

> {{% admonition type="note" %}}
> This component requires [Role-based access control (RBAC)][] to be set up
> in Kubernetes for {{< param "PRODUCT_ROOT_NAME" >}} to access it via the Kubernetes REST API.
> 
> Role-based access control (RBAC)]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/
> {{% /admonition %}}

Reviewer: ```suggestion
{{< admonition type="note" >}}
This component requires [Role-based access control (RBAC)][] to be set up
in Kubernetes for {{< param "PRODUCT_ROOT_NAME" >}} to access it via the Kubernetes REST API.

Role-based access control (RBAC)]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/
{{< /admonition >}}
```

Minor shortcode syntax update
